Add encryption for sensitive submitted data

NHAS opened the pull request:
Hi, I've added the ability to encrypt sensitive data submitted to gophish from those who have been phished.
This allows for the phisher to add public keys to gophish and then select them in the campaign creation menu.

Commit:
Add current work for adding encryption

Commit:
Adding public keys to database now correctly works

Starting on rewriting parts of the API

Add the encryption to specific events

Add public key model

Update campaigns to allow for encryption selection

Add tests for encryption and use api properly

Update mysql migration file

Commit:
Standardise user interaction to match rest of gophish and implement all api calls for public key

Change error messages to match public key

Fix bad path

Fix small gui bug

Return script to minified form

Make public key selection in campaign

Commit:
Add fix for put requests being able to generate duplicate name entries

Commit:
Disallow removal of public if in use

Minify sources

Commit:
Add campaign results information for encrypted data

Commit:
Cleanup ids

jordan-wright commented on auth/auth.go:
```diff
@@ -97,6 +98,10 @@ func Register(r *http.Request) (bool, error) {
 	u.Hash = string(h)
 	u.ApiKey = GenerateSecureKey()
 	err = models.PutUser(&u)
+	if err != nil {
```
Let's leave out changes to this file, since the only addition is an error check. We can keep things scoped to just the changes being made.

jordan-wright commented on db/db_mysql/migrations/20160118194630_init.sql:
```diff
@@ -2,6 +2,7 @@
 -- +goose Up
 -- SQL in section 'Up' is executed when this migration is applied
 CREATE TABLE IF NOT EXISTS users (id integer primary key auto_increment,username varchar(255) NOT NULL UNIQUE,hash varchar(255),api_key varchar(255) NOT NULL UNIQUE );
+CREATE TABLE IF NOT EXISTS public_keys (id integer primary key autoincrement, user_id integer NOT NULL, friendly_name varchar(255), pub_key blob NOT NULL);
```
Let's create a new database migration for this instead of editing old ones. This makes it easier to keep track of changes that have been applied over time.

jordan-wright commented on db/db_mysql/migrations/20160131153104_0.1.2_add_event_details.sql:
```diff
@@ -2,6 +2,7 @@
 -- +goose Up
 -- SQL in section 'Up' is executed when this migration is applied
 ALTER TABLE events ADD COLUMN details BLOB;
+ALTER TABLE events ADD COLUMN key BLOB;
```
Likewise here, let's move this into a separate migration for these changes.

jordan-wright commented on models/group.go:
```diff
@@ -212,6 +213,12 @@ func PutGroup(g *Group) error {
 	}
+
+	record, err := GetGroupByName(g.Name, g.UserId)
+	if err == nil && record.Id != g.Id {
```
Let's remove the checks to this new error you're introducing. I see the reasoning behind it, but it's better to keep these changes focused solely on the public key support you're adding.

NHAS replied:
I believe this issue isnt from my input, what I found during development of this is that put requests can modify the names of landing pages/groups/etc and set them to be the same name as other entries. 
However you are correct in that this should be a seperate merge on its own, rather than being lumped in with the public key stuff.

jordan-wright commented on models/page.go:
```diff
@@ -138,6 +138,16 @@ func PostPage(p *Page) error {
+		return err
+	}
+
+	record, err := GetGroupByName(p.Name, p.UserId)
```
As mentioned earlier, let's remove all the added checks in here, since it increases the scope of these changes. Also, I'm not quite sure what you're doing here, since you're getting a group when the user is editing a landing page?

jordan-wright commented on models/result.go:
```diff
@@ -49,7 +55,49 @@ func (r *Result) createEvent(status string, details interface{}) (*Event, error)
 		}
-		e.Details = string(dj)
+
+		if EVENT_DATA_SUBMIT == status && c.PublicKeyId != 0 { // Zero is unset
```
Minor nit - let's switch this around
```suggestion
		if status == EVENT_DATA_SUBMIT && c.PublicKeyId != 0 { // Zero is unset
```

jordan-wright commented on models/result.go:
```diff
@@ -49,7 +55,49 @@ func (r *Result) createEvent(status string, details interface{}) (*Event, error)
-		e.Details = string(dj)
+
+		if EVENT_DATA_SUBMIT == status && c.PublicKeyId != 0 { // Zero is unset
+			//Taken from crypto/cipher CFB example
```
I'll have to review this a bit more closely. Tbh, I'm not a crypto expert so I'll take a look at the examples provided in the Golang source that you mentioned.

jordan-wright replied:
Also, let's move this to a function in `publickey.go`. Perhaps creating a function like this (untested):
```suggestion
			func (pk *PublicKey) Encrypt(data []byte) ([]byte, error) {} 
```

NHAS replied:
That's a good suggestion. As for the crypto side of things, if you'd like me to do a bit of a walk through of my thought process for it I'd be more than happy. 
I would suggest reading through the source first, so you spot anything I might bias you against with an explanation.

jordan-wright commented on models/smtp.go:
```diff
@@ -203,6 +204,12 @@ func PutSMTP(s *SMTP) error {
 		return err
 	}
+
+	record, err := GetSMTPByName(s.Name, s.UserId)
```
As mentioned before, let's remove these checks that aren't necessary for these changes.

jordan-wright commented on models/template.go:
```diff
@@ -137,6 +137,12 @@ func PutTemplate(t *Template) error {
 		return err
 	}
+
+	record, err := GetGroupByName(t.Name, t.UserId)
```
As mentioned before, let's remove these checks that aren't necessary for these changes.

jordan-wright commented on models/user.go:
```diff
@@ -35,5 +35,9 @@ func GetUserByUsername(username string) (User, error) {
 // PutUser updates the given user
 func PutUser(u *User) error {
 	err := db.Save(u).Error
+	if err != nil {
```
As mentioned before, let's remove these checks that aren't necessary for these changes.

jordan-wright commented:
Hi @NHAS,

Thanks for sending this in! This is quite a large pull request, so it's going to take some time to work through. I've added some notes to get started.

But, at a higher level, what benefits do you see here besides being able to encrypt submitted data? Are there any other uses for these public keys that you can envision (it's ok if there's not).

Also, what are your thoughts around making decryption as easy as possible for users? Unless I'm missing something, it appears that there would be quite a bit of friction right now for users trying to decrypt submitted data. They would have to export the data and manually figure out how to use their private key to decrypt the data, and that kind of friction goes against the UX we've strived for so far.

I'm open to ideas, but we'll definitely want to figure out the decryption story before getting this exposed to everyone.

That said, this is a really interesting concept and I appreciate all the hard work you put into the pull request! You definitely took some time to make sure the bases were covered, which is always appreciated.

NHAS commented:
Hi, thanks for your quick response!

Currently I can't think of any extension for the public keys, other than being able to have different levels of paranoia for the encryption. 
I.e 
None = No public key, nothing encrypted.
Low = Just sensitive submitted data encrypted.
High = All data submitted is encrypted.

I've made a small command line tool that takes the exported raw results csv file and decrypts it given the public key. I can upload that to git if you want to have a look at it, although its not very polished. 
My personal bias is to not have the decryption done on the server side at all. 
However if it needs to be done for UX reasons I'd suggest having a box popup, when exporting anything with encrypted results, where the user can copy and paste their private RSA key. This would then be used for that decryption, and promptly flushed.

As Im not particularly familiar with forking in git, would you prefer me to add new commits to my fork and remerge?
Or is it better to go back and rewrite my commits?

jordan-wright committed:
Update models/result.go

Co-Authored-By: NHAS <NHAS@users.noreply.github.com>

Commit:
Move encryption behaviour to publickey model

Commit:
Revert PUT request fix

Commit:
Create new migration files

Commit:
Add mitigation files

Commit:
Revert mysql database file

Commit:
Merge master changes

jordan-wright commented:
Hey @NHAS,

I see you've been adding commits to this pull request. Thank you for your continued work on this!

Let me know when you're ready for another code review, and I'll take a look at things. I didn't want to review things if you were still actively working on them.

Commit:
Merge gophish master

Commit:
Add ability to decrypt events before downloading

Commit:
Remove logging of public key

NHAS commented:
I've just added the ability to decrypt events before downloading them, I'd say this meets your requested changes.
As such, I'd say now would be a good time for a code review.

TheGroundZero commented:
May I recommend adding an option to hash sensitive information using SHA1?
I'm thinking about running captured passwords thorugh [Have I Been Pwned](https://www.troyhunt.com/ive-just-launched-pwned-passwords-version-2/#cloudflareprivacyandkanonymity), which can be queried using SHA1-hashed passwords?

NHAS commented:
@TheGroundZero 
I agree that it would add some confidentiality to the passwords and is probably a good idea. 
The only problem with this is the fact that they would still need to be encrypted on the server, due to the fact unsalted SHA1 is unsuitable for storing passwords. 

To quote troy hunt:
"It doesn't matter that **SHA1 is a fast algorithm unsuitable for storing your customers' passwords** with because that's not what we're doing here, it's simply about ensuring the source passwords are not immediately visible."

TheGroundZero commented:
> @TheGroundZero
> I agree that it would add some confidentiality to the passwords and is probably a good idea.
> The only problem with this is the fact that they would still need to be encrypted on the server, due to the fact unsalted SHA1 is unsuitable for storing passwords.

My view is that you'd only capture sensitive data if this is approved by your client, your employer, ... and that you wouldn't store this data for a long time period (plus the system running GoPhish would be properly secured).

The use of SHA1 would add a (small) barrier to being able to view the data in plain-text, which would mean someone would need to actively perform an action to retrieve the original data, breaking any contract there would be with a client.

I would of course encourage the secure storage of any sensitive data on a system, including the GoPhish database.  
You could even combine an encrypted storage solution with (optionally) storing the sensitive data in a hashed format.

NHAS commented:
@TheGroundZero 
I definitely think this should be another pull request. But perhaps after this one gets merged, as I already have encryption working for plaintext credentials.


Then it would be, say a checkbox in the encryption settings.

NHAS committed:
Update public key api to new api structure

NHAS committed:
Add public key api file

NHAS commented:
@jordan-wright 
I've just updated the encryption to use the new api scheme. If you wanted to have a look at the code and give some feedback, that'd be great!

rmarks-jacobs commented:
Had this exact same idea tonight and asked one of my team members to look into whether we could use AWS CloudFront's support for field level encryption to satisfy this requirement as a stop-gap solution, but looks like you are close to having this native support added.  Thanks for the hard work!

NHAS commented:
@rmarks-jacobs 
Thanks!
If anyone on your team has experience with encryption and whatnot, it'd be good to get some feedback on what I've done.

jordan-wright committed:
Move API key to Bearer Token (#1439)

* Moved api_key from URL to authorization header in requests (#1434)

* Fixing some minor formatting and rebuilding minified JS

jordan-wright committed:
Update Javascript Dependencies (#1440)

* updated devDependencies, migrated gulpfile.js to gulp 4.0 syntax (#1438)

* Rebuilding JS dependencies with new gulp config. Updated yarn.lock.

Co-authored-by: Christian Schwartz <christian.schwartz@gmail.com>

aczietlow committed:
Update Sending Profile Message-ID headers (#1417) (#1441)

Adds a default message-ID header to outbound emails.

wjwoodson committed:
Add Dockerfile (#1437)

Added dockerfile

Sevalel commented:
Hey i am planning to run a preventive phishing campaign in the hospital in which im working. But i have no right to know their password according to the headoffice (I guess they're right). Is there any way to encrypt these even while checking the "capture password" box?
I mean have instead of lets say for exemple "WHATEVER" as a password, somthing like WH****ER, in the "submitted datas" section?

TheGroundZero commented:
> 
> 
> Hey i am planning to run a preventive phishing campaign in the hospital in which im working. But i have no right to know their password according to the headoffice (I guess they're right). Is there any way to encrypt these even while checking the "capture password" box?
> I mean have instead of lets say for exemple "WHATEVER" as a password, somthing like WH****ER, in the "submitted data" section?

If you have no business with the password, I wouldn't know why you'd capture them in the first place.
Displaying passwords plain-text, even partially may/will have an impact on the confidentiality of the password.
Would your idea be to show the partial password to the user and ask whether this was in fact his/her password? If so, that would lead to a whole different discussion of revealing the identity of those who've fallen for the phish.

However, this whole discussion is taking this Pull Request off-topic.
The original PR by NHAS was to include encrypted storage of the captured data on the GoPhish server for security purposes.
My addition to that was to hash the passwords, so you can test them against HIBP or rainbow tables, without reading any clear-text password, adding confidentiality. Which would probably also be better served with a separate PR.

Sevalel commented:
Sorry about that, just delete this discussion right after if it's off-topic.

The aim of this campaign is to educate internet users in the hospital about phishing and the risks of such an activity. We want to show them that we can easily get their logs because they’re not suspicious enough and then explain them how to be more careful while using internet. 

But at the same time we do not want to know their password for confidentiality reasons, just the first one or two letters for example instead of the full clear-text password in "submitted data", to show them that a real hacker can get more. And they will be invited to change their password if they want to. 

Just wondering if such a thing was doable using the v0.7.1.

dmaciejak committed:
Prevent adding an empty recipient (#1448)

gliptak committed:
Add Go 1.12 to Travis (#1472)

jordan-wright committed:
Implement User Management API (#1473)

This implements the first pass for a user management API allowing users with the `ModifySystem` permission to create, modify, and delete users. In addition to this, any user is able to use the API to view or modify their own account information.

jordan-wright committed:
Added a check to ensure the target details are correct if manually created.

Fixes #1475

cschwartz committed:
Add support for encrypted connections to mysql (#1460)

dependabot[bot] committed:
Bump lodash.merge from 4.6.1 to 4.6.2 (#1508)

Bumps [lodash.merge](https://github.com/lodash/lodash) from 4.6.1 to 4.6.2.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/commits)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump lodash from 4.17.11 to 4.17.14 (#1507)

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.11 to 4.17.14.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.11...4.17.14)

Signed-off-by: dependabot[bot] <support@github.com>

jordan-wright committed:
Add CORS support for Reporting Handler (#1529)

* Added response headers for CORS and server identification (#1517)

Co-Authored-By: Glenn Wilkinson <glenn.wilkinson@gmail.com>

dmaciejak committed:
Fix multiple XSS issues in User Management Page (#1547)

If the user name is embedding some JS code, it will be executed on the client side. Note: gophish/static/js/dist/app/users.min.js will need to be regenerated too.

dmaciejak committed:
Remove the Top Navigation Menu

The top navigation menu is redundant to the sidebar. We’re opting to remove it so that we are free to update the sidebar to be more complex over time.

NHAS committed:
Merge with new upstream code

NHAS committed:
Fix small issue misidentifying which code to merge

jordan-wright committed:
Upgrade SweetAlert2 Dependency (#1583)

Upgrades the SweetAlert2 dependency to version 8.x.x.

Co-authored-by: Glenn Wilkinson <glenn.wilkinson@gmail.com>

gliptak committed:
Add Go 1.13 to Travis (#1610)

quelsan committed:
Improved Dockerfile and run script (#1579)

The Dockerfile has also been rebuilt to use an unprivileged user instead
of root.

The run script adds more options and changes the method the
configuration is overwritten, which may help in situations where the
configuration is not owned by the unprivileged user.

glennzw committed:
Will exit on port binding failure (#1635)

SamSutt committed:
Bugfix Issue 1570 (#1640)

Relating to the Issue (https://github.com/gophish/gophish/issues/1570), this could be an easy fix.

dependabot[bot] committed:
Bump mixin-deep from 1.3.1 to 1.3.2 (#1654)

Bumps [mixin-deep](https://github.com/jonschlinkert/mixin-deep) from 1.3.1 to 1.3.2.
- [Release notes](https://github.com/jonschlinkert/mixin-deep/releases)
- [Commits](https://github.com/jonschlinkert/mixin-deep/compare/1.3.1...1.3.2)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump lodash from 4.17.11 to 4.17.15 (#1509)

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.11 to 4.17.15.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.11...4.17.15)

Signed-off-by: dependabot[bot] <support@github.com>

Commit:
1624 - Return API Key on Password Change (#1663)

Updates the user model to return the user’s API key for easier administration.

jordan-wright committed:
Updated the Dockerfile to allow the gophish binary to bind to privileged ports. Fixes #1660.

jordan-wright committed:
Performance Improvements for Campaign and Group Creation (#1686)

This commit significantly improves the performance of campaign and group creation by changing database access to use transactions.

It should also make things more consistent with campaign creation. Specifically, this will ensure that the entire campaign gets created before emails start sending, while I anticipate this will fix #1643, #1080, (possibly) #1677, and #1552.

jordan-wright committed:
Updates the tls.Config of the phishing and admin servers to support TLS 1.2 as the minimum TLS version. This addresses #1691 and #1689.

I am making this change since Microsoft, Google, and Apple have all chosen to deprecate TLS 1.0 and TLS 1.1 in early 2020. In late 2018, the companies recorded that less than 1.4 percent (max) of their connections used < TLS 1.2.

Output before change:

```
docker run --rm -ti -p 3333:3333 drwetter/testssl.sh https://host.docker.internal:3333

 Testing protocols via sockets except NPN+ALPN

 SSLv2      not offered (OK)
 SSLv3      not offered (OK)
 TLS 1      offered (deprecated)
 TLS 1.1    offered (deprecated)
 TLS 1.2    offered (OK)
 TLS 1.3    offered (OK): final
 NPN/SPDY   h2, http/1.1 (advertised)
 ALPN/HTTP2 h2, http/1.1 (offered)
```

Output after change:

```
docker run --rm -ti -p 3333:3333 drwetter/testssl.sh https://host.docker.internal:3333

 Testing protocols via sockets except NPN+ALPN

 SSLv2      not offered (OK)
 SSLv3      not offered (OK)
 TLS 1      not offered
 TLS 1.1    not offered
 TLS 1.2    offered (OK)
 TLS 1.3    offered (OK): final
 NPN/SPDY   h2, http/1.1 (advertised)
 ALPN/HTTP2 h2, http/1.1 (offered)
```

jordan-wright committed:
Remove support for Go 1.9.

It appears that the MySQL driver dropped support for Go 1.9. Until we have vendoring in place (will happen soon!) we must follow suit and drop support for Go v1.9. This still allows us to support the latest three versions of Go.

Commit:
Add Webhook Support

Adds support for managing outgoing webhooks. Closes #1602

jordan-wright committed:
Minor cleanup on webhook feature integration

- Ran gofmt
- Rebuilt minified static files
- Updated validation payload

jordan-wright committed:
Added gofmt requirement to CI

Commit:
Update javascript and fork with newest changes

NHAS committed:
Update verson number

NHAS commented:
Is there anything more I can do to help with this merge request? @jordan-wright

aczietlow committed:
Fix webhooks db migration. (#1697)

jordan-wright committed:
Deleted unneeded .DS_Store file that was accidentally merged as part of #1642.

JoseW3b committed:
Get send_by_date (#1720)

Get send_by_date field when query on GetCampaignSummary and GetCampaignSummaries

jordan-wright committed:
Removing unneeded query to improve group performance.

As mentioned in #1702, the query in `insertTargetIntoGroup` isn't needed, since both instances where it's used we already know that the target isn't in the group. This means it's safe to remove that query, improving performance dramatically.

Before:

```
BenchmarkPostGroup100-4     	      81	  12629754 ns/op	 2989993 B/op	   52918 allocs/op
BenchmarkPostGroup1000-4    	       6	 189527792 ns/op	29891818 B/op	  528082 allocs/op
BenchmarkPostGroup10000-4   	       1	6203645806 ns/op	299253648 B/op	 5282859 allocs/op
BenchmarkPutGroup100-4      	     100	  10221833 ns/op	 2589165 B/op	   46078 allocs/op
BenchmarkPutGroup1000-4     	       7	 162692432 ns/op	25812440 B/op	  458425 allocs/op
BenchmarkPutGroup10000-4    	       1	7596445084 ns/op	260385808 B/op	 4581569 allocs/op
```

After:

```
BenchmarkPostGroup100-4     	     133	   8848973 ns/op	 2354605 B/op	   42322 allocs/op
BenchmarkPostGroup1000-4    	       9	 112557410 ns/op	23549206 B/op	  422184 allocs/op
BenchmarkPostGroup10000-4   	       1	3414209403 ns/op	235635952 B/op	 4222090 allocs/op
BenchmarkPutGroup100-4      	     147	   8094333 ns/op	 2271297 B/op	   40777 allocs/op
BenchmarkPutGroup1000-4     	       9	 125092124 ns/op	22635067 B/op	  405421 allocs/op
BenchmarkPutGroup10000-4    	       1	5712591900 ns/op	228592920 B/op	 4051316 allocs/op
```

jordan-wright committed:
Refactoring Logging (#1722)

* Added ParseLevel to set log level (#1671)
* Moved logger config into the logger package for better decoupling. Added logging tests.

Co-authored-by: Amal Alkhamees <Amalkh5@users.noreply.github.com>

glennzw committed:
Added IMAP support for checking reported emails (#1612)

Initial support of managing reporting through IMAP.

Co-Authored-By: Jordan Wright <jmwright798@gmail.com>

jordan-wright committed:
Rebuilding minified JS to support #1722

jordan-wright committed:
Bumping version to 0.9.0

jordan-wright committed:
Adjusting SMTP TLS config to use just the hostname instead of the hostname+port when validating certificates. Fixes #1709

jordan-wright committed:
Adding benchmarks for maillog.Generate. Ref #1726

jordan-wright committed:
Added campaign ID to the webhook JSON payload

J3rry367 commented:
Are there any updates on this? Having the ability to have the credentials encrypted would be a huge help. I just came across this project and am thinking of implementing it, but it will be a big hurdle if the credentials cannot be encrypted. The way we currently do phishing is the submitted details GPG encrypted and emailed to the person/people running the campaign in order to allow them to weed out anyone who filled in blank creds and/or attempted to hack the phishing page during the campaign. As I understand if you do not capture the data submitted the target may have just clicked the submit button without filling out the form. I'd like to filter those out without compromising on security any more than necessary.

cyberbutler committed:
Updated Docker Hub link to point to gophish/gophish (#1732)

The Readme currently points to https://hub.docker.com/r/matteoggl/gophish/ which is a Docker image that hasn't been updated in over 2 years. This PR updates the Readme link to point to a more current Docker Image.

jordan-wright committed:
Refactoring tests to remove stretchr/testify dependency

jordan-wright committed:
Create ci.yml (#1741)

This PR enabled GitHub Actions as a replacement to TravisCI. I'm a fan of removing external dependencies where possible, and GitHub Actions has matured to be really useful.

NHAS commented:
As for implementation this PR has the functionality of being able to provide a public key that the server will then encrypt all results with. Blank or not. 

I think the functionality for filtering out blank results doesn't fit the scope of this PR to add encryption. After downloading the results of a phishing campaign you could write a tool to filter them before sending I'd imagine.

jordan-wright committed:
Updated README to include GitHub Actions badge and update LICENSE copyright date

J3rry367 commented:
Yeah, I get that. I was just saying there needs to be a way to tell if the submissions are blank/faked which you can only do if you capture the info, but you don't particularly want to capture the info if it's not encrypted. Mostly just illustrating the need for this to be implemented.

jordan-wright committed:
Creating go.mod and go.sum files to establish Gophish as a Go module.

NHAS committed:
Fix merge issues

NHAS committed:
Fix issues with UI, when deleting keys associated with campaigns

l2k-ais commented:
Are there any more blockers for this @jordan-wright ? I'd love to see this merged.

NHAS commented:
Unfortunately I imagine my fork is now quite out of date. If there is still interest from @jordan-wright in this Im happy to update it in the coming months. 
But otherwise I'll probably close this as it hasnt gone anywhere.

NHAS commented:
Well, Im going to close this pull request as nothing has happen and nothing will probably happen around this.
